compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/ContextMenuProvider.desktop.kt

```diff
@@ -99,6 +101,19 @@ private val LocalContextMenuData = staticCompositionLocalOf<ContextMenuData?> {
 }
 
 
+/**
+ * Returns whether the given event is a context menu opening gesture.
+ */
+private fun PointerEvent.isOpenContextMenuGesture(): Boolean {
```

Comment:
We have this logic on the interop level (search for `isMacOsCtrlClick`).

If we are sure that it should be only for context menus, we should remove it from there.

Reply:
Oh, I didn't know that.

But I think we shouldn't be reporting ctrl-clicks as right-clicks in such a low-level API as `PointerButtons`. I suggest we add a (public) `PointerEvent.isPopupTrigger`, and have `Modifier.contextMenuOpenDetector` use it.

Reply:
> But I think we shouldn't be reporting ctrl-clicks as right-clicks in such a low-level API

I also think that, but before doing the changes, better investigate how the native API behaves.

> PointerEvent.isPopupTrigger, and have Modifier.contextMenuOpenDetector

Do we need `isPopupTrigger`, if we already have `contextMenuOpenDetector` as public?
- I am not sure that it is needed for popups other than context menus
- `contextMenuOpenDetector` hides gesture detection under the hood, not reacting on a single event. It can hide consuming logic, or event sequence logic (first check for down, and only then for up)

Reply:
> I also think that, but before doing the changes, better investigate how the native API behaves.

What do you mean? AWT reports the value in `MouseEvent.isPopupTrigger`. It doesn't report ctrl-clicks as right-clicks.

> Do we need isPopupTrigger, if we already have contextMenuOpenDetector as public?

Well, we need it anyway in order to correctly implement `contextMenuOpenDetector` if we remove the reporting of ctrl-clicks as right-clicks. The question is whether we should make it public.
I think we should, because users may want to react to that event in places unrelated to context menus. They may not even think to look in `DropdownMenu.desktop.kt` (in the material lib!).
Also, as you say, `contextMenuOpenDetector` can (and should) have additional logic, which the user may not need.

Reply:
> What do you mean? AWT reports the value in MouseEvent.isPopupTrigger. It doesn't report ctrl-clicks as right-clicks.

I mean the native API on Swift/Objective C, not AWT
 - does it provide a low-level API that don't react on ctrl
 - does it provide a high-level API that reacts on ctrl?

Reply:
> may want to react to that event in places unrelated to context menus. 

The question - why they need to react on "popup" trigger for non-context menus. If there are cases, we can think how can we cover them. If we won't find cases, then exposing it as public isn't useful I think, it adds burden of supporting /explaining it.

> material lib

`contextMenuOpenDetector` is in foundation.

> have additional logic, which the user may not need.

If they need own logic, they can write their own function. The library shouldn't provide everything, it should provide things that cover useful use cases.

Reply:
> The question - why they need to react on "popup" trigger for non-context menus. If there are cases, we can think how can we cover them.

Maybe they want to implement their own popups, maybe in a game or something. Guessing what the user wants is always hard, and I thought you were usually in favor of providing him with the tools to do anything.

> contextMenuOpenDetector is in foundation.

Right, I was looking at the private one in `DropdownMenu.desktop.kt`.

> If they need own logic, they can write their own function. The library shouldn't provide everything, it should provide things that cover useful use cases.

They will need to do `(event.nativeEvent as MouseEvent).isPopupTrigger`, which is not a good thing.

I'll check native APIs, but I think the fact that Java provides it as a property of `MouseEvent` is already a strong point in favor.

Reply:
> Maybe they want to implement their own popups, maybe in a game or something.

But this popup will be called a context menu. Or maybe context popup 🤔. The logic we are trying to provide is to open something that is "context", and it has different triggers on different systems. These triggers can be not a simple events, but a series of events. And this series of events might be impossible to describe as an extension of a single event (though, we can with the current API and platforms)

> Java provides it as a property of MouseEvent

The Java API designed for Java event system which doesn't have gestures. Besides, this API is old and controversial.

Reply:
> impossible to describe as an extension of a single event (though, we can with the current API and platforms)

- If we wouldn't have key modifiers in PointerEvent, we have to introduce "input gesture" and listen both key and mouse
- if a platform don't simple open a context menu on a Release, and listens for a gesture "press, check that we don't move, release", we would have to listen 2 events and check the delta between them.

Reply:
As an alternative, I suggest to provide `suspend PointerInputScope.detectContextMenuOpener`